pkg/kapp/clusterapply/cluster_change_set.go

```diff
@@ -95,18 +99,30 @@ func (c ClusterChangeSet) Apply(changesGraph *ctldgraph.ChangeGraph) error {
 
 	blockedChanges := ctldgraph.NewBlockedChanges(changesGraph)
 	applyingChanges := NewApplyingChanges(
-		expectedNumChanges, c.opts.ApplyingChangesOpts, c.clusterChangeFactory, c.ui)
-	waitingChanges := NewWaitingChanges(expectedNumChanges, c.opts.WaitingChangesOpts, c.ui)
+		expectedNumChanges, c.opts.ApplyingChangesOpts, c.clusterChangeFactory, c.ui, c.opts.ExitOnError)
+	waitingChanges := NewWaitingChanges(expectedNumChanges, c.opts.WaitingChangesOpts, c.ui, c.opts.ExitOnError)
+
+	var unsuccessfulChanges []string
 
 	for {
-		appliedChanges, err := applyingChanges.Apply(blockedChanges.Unblocked())
+		appliedChanges, unsuccessfulChangeDesc, err := applyingChanges.Apply(blockedChanges.Unblocked())
 		if err != nil {
 			return err
 		}
 
+		unsuccessfulChanges = append(unsuccessfulChanges, unsuccessfulChangeDesc...)
+
 		waitingChanges.Track(appliedChanges)
 
 		if waitingChanges.IsEmpty() {
+			if len(unsuccessfulChanges) == 1 {
```

Comment:
`NewSemiStructuredError` doesn't work really well when there is only one error inside. `[]` because there might be errors containing something like `map[string]string` and there is no easy way to differentiate between an error and such field.
I should probably add this as a comment.